src/openzeppelin/token/erc721.cairo

```diff
@@ -0,0 +1,354 @@
+        Approval(owner, to, token_id);
+    }
+
+    #[internal]
+    fn _set_approval_for_all(owner: ContractAddress, operator: ContractAddress, approved: bool) {
+        assert(owner != operator, 'ERC721: self approval');
+        _operator_approvals::write((owner, operator), approved);
+        ApprovalForAll(owner, operator, approved);
+    }
+
+    #[internal]
+    fn _mint(to: ContractAddress, token_id: u256) {
+        assert(_owner_of(token_id).is_none(), 'ERC721: token already minted');
+        assert(!to.is_zero(), 'ERC721: invalid receiver');
+
+        // Update balances
+        _balances::write(to, _balances::read(to) + u256_from_felt252(1));
+
+        // Update token_id owner
+        _owners::write(token_id, to);
+
+        // Emit event
+        Transfer(contract_address_const::<0>(), to, token_id);
+    }
+
+    #[internal]
+    fn _transfer(from: ContractAddress, to: ContractAddress, token_id: u256) {
+        let owner = _owner_of(token_id).expect('ERC721: invalid token ID');
+        assert(from == owner, 'ERC721: wrong sender');
+        assert(!to.is_zero(), 'ERC721: invalid receiver');
+
+        // Implicit clear approvals, no need to emit an event
+        _token_approvals::write(token_id, contract_address_const::<0>());
+
+        // Update balances
+        _balances::write(from, _balances::read(from) - u256_from_felt252(1));
+        _balances::write(to, _balances::read(to) + u256_from_felt252(1));
+
+        // Update token_id owner
+        _owners::write(token_id, to);
+
+        // Emit event
+        Transfer(from, to, token_id);
+    }
+
+    #[internal]
+    fn _burn(token_id: u256) {
+        let owner = _owner_of(token_id).expect('ERC721: invalid token ID');
+
+        // Implicit clear approvals, no need to emit an event
+        _token_approvals::write(token_id, contract_address_const::<0>());
+
+        // Update balances
+        _balances::write(owner, _balances::read(owner) - u256_from_felt252(1));
+
+        // Delete owner
+        _owners::write(token_id, contract_address_const::<0>());
+
+        // Emit event
+        Transfer(owner, contract_address_const::<0>(), token_id);
+    }
+
+    #[internal]
+    fn _safe_mint(to: ContractAddress, token_id: u256, data: Array<felt252>) {
+        _mint(to, token_id);
+        assert(
+            _check_on_erc721_received(contract_address_const::<0>(), to, token_id, data),
+            'ERC721: safe mint failed'
+        );
+    }
+
+    #[internal]
+    fn _safe_transfer(
+        from: ContractAddress, to: ContractAddress, token_id: u256, data: Array<felt252>
+    ) {
+        _transfer(from, to, token_id);
+        assert(_check_on_erc721_received(from, to, token_id, data), 'ERC721: safe transfer failed');
+    }
+
+    #[internal]
+    fn _set_token_uri(token_id: u256, token_uri: felt252) {
+        assert(_exists(token_id), 'ERC721: invalid token ID');
+        _token_uri::write(token_id, token_uri)
+    }
```

Comment:
I followed [the implementation in the main branch](https://github.com/OpenZeppelin/cairo-contracts/blob/main/src/openzeppelin/token/erc721/library.cairo#L68), which tracks the ERC721StorageURI approach.

Fix uri works well for ERC1155 because of this specs

> Metadata
> 
> The URI value allows for ID substitution by clients. If the string {id} exists in any URI, clients MUST replace this with the actual token ID in hexadecimal form. This allows for a large number of tokens to use the same on-chain string by defining a URI once, for that large number of tokens.
> 
>     The string format of the substituted hexadecimal ID MUST be lowercase alphanumeric: [0-9a-f] with no 0x prefix.
>     The string format of the substituted hexadecimal ID MUST be leading zero padded to 64 hex characters length if necessary.
> 
> Example of such a URI: https://token-cdn-domain/{id}.json would be replaced with https://token-cdn-domain/000000000000000000000000000000000000000000000000000000000004cce0.json if the client is referring to token ID 314592/0x4CCE0.

Unfortunately, there is not such substitution standardized in ERC721, so all tokens would get the same uri, and thus the same metadata.

Open to discussing what we want to do. A possibility is to store a "prefix" and to have `token_uri(u256)` return the concatenation of the prefix and the stringified token ID